src/include/duckdb/function/udf_function.hpp

```diff
@@ -0,0 +1,317 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/function/udf_function.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/common/types.hpp"
+#include "duckdb/function/scalar_function.hpp"
+#include "duckdb/main/client_context.hpp"
+#include "duckdb/parser/parsed_data/create_scalar_function_info.hpp"
+#include "duckdb/common/types/data_chunk.hpp"
+#include <string>
+#include <vector>
+#include <type_traits>
+#include<tuple>
+
+using namespace std;
+
+namespace duckdb {
+
+class UDFWrapper {
+public:
+	UDFWrapper(ClientContext &context);
+
+	template<typename TR, typename... Args> void CreateFunction(string name, TR (*udf_func)(Args...)) {
+		const std::size_t num_template_argc = sizeof...(Args);
+		switch(num_template_argc) {
+			case 1:
+				CreateUnaryFunction<TR, Args...>(name, udf_func);
+				break;
+			case 2:
+				CreateBinaryFunction<TR, Args...>(name, udf_func);
+				break;
+			case 3:
+				CreateTernaryFunction<TR, Args...>(name, udf_func);
+				break;
+			default:
+				throw duckdb::Exception(ExceptionType::EXECUTOR, "UDF function only supported until ternary!");
+		}
+	}
+
+	template<typename TR, typename... Args>
+	void CreateFunction(string name, vector<SQLType> args, SQLType ret_type, TR (*udf_func)(Args...)) {
+		if(!TypesMatch<TR>(ret_type)) {
+			string msg("Return type doesn't match with the first template type: ");
+			msg += typeid(TR).name();
+			msg += ".";
+			throw duckdb::Exception(ExceptionType::MISMATCH_TYPE, msg);
```

Comment:
This doesn't follow the general way of constructing DuckDB exceptions, in which the formatting is done in the Exception constructor. This should be something like `throw duckdb::ParserException("Return type doesn't match with the first template type: %s.", typeid(TR).name().c_str())`.

We also already have an exception for this: `TypeMismatchException`. I would change it to throw that exception instead.